File: c/problem_64.c
#include "problem_64.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

static int board64[8][8];
static int moves64[8][2] = {{2,1},{2,-1},{-2,1},{-2,-1},{1,2},{1,-2},{-1,2},{-1,-2}};
static int n64;

static void dfs64(int r, int c, int visited, int *count) {
    if (visited == n64*n64) { (*count)++; return; }
    for (int m=0;m<8;m++) {
        int nr=r+moves64[m][0], nc=c+moves64[m][1];
        if (nr>=0&&nr<n64&&nc>=0&&nc<n64&&!board64[nr][nc]) {
            board64[nr][nc]=1; dfs64(nr,nc,visited+1,count); board64[nr][nc]=0;
        }
    }
}

int knight_tours(int n) {
    n64=n; int count=0;
    memset(board64,0,sizeof(board64));
    for(int r=0;r<n;r++) for(int c=0;c<n;c++) {
        board64[r][c]=1; dfs64(r,c,1,&count); board64[r][c]=0;
    }
    return count;
}

void problem64(void) {
    printf("Problem 64:\n");
    assert(knight_tours(1)==1);
    assert(knight_tours(2)==0);
    assert(knight_tours(3)==0);
    printf("All Tests Passed\n");
}

File: typescript/problem_64.ts
export function knightTours(n: number): number {
    const board: boolean[][] = Array.from({length:n}, () => new Array(n).fill(false));
    const moves = [[2,1],[2,-1],[-2,1],[-2,-1],[1,2],[1,-2],[-1,2],[-1,-2]];
    let count = 0;
    function dfs(r: number, c: number, visited: number): void {
        if (visited === n*n) { count++; return; }
        for (const [dr,dc] of moves) {
            const nr=r+dr, nc=c+dc;
            if (nr>=0&&nr<n&&nc>=0&&nc<n&&!board[nr][nc]) {
                board[nr][nc]=true; dfs(nr,nc,visited+1); board[nr][nc]=false;
            }
        }
    }
    for (let r=0;r<n;r++) for (let c=0;c<n;c++) { board[r][c]=true; dfs(r,c,1); board[r][c]=false; }
    return count;
}

export function problem64(): void {
    console.log("Problem 64:");
    if (knightTours(1) !== 1) throw new Error("assertion failed");
    if (knightTours(2) !== 0) throw new Error("assertion failed");
    if (knightTours(3) !== 0) throw new Error("assertion failed");
    console.log("All Tests Passed");
}
